Add Heuristic Region in Cold Block Marker

dsouzai opened the pull request:
The getClassFromSignature call in TR_ColdBlockMarker::hasNotYetRun is
only used to make a heuristic decision as to whether or not the block
should be marked cold. In a relocatable compile (currently in OpenJ9),
the getClassFromSignature will result in a validation record that will
be used in the load run. This validation could fail because the class
from signature might not yet exist. However, executing the body in a
subsequent run does not require this class to exist (assuming this is
the only reference to it during the compilation), since its existence
was only used for a heuristic decision regarding cold block marking -
an act that only mattered during the compilation.

This PR wraps the getClassFromSignature call in a heuristic
region.

Signed-off-by: Irwin D'Souza <dsouzai.gh@gmail.com>

0xdaryl commented:
The heuristic region logic does not live in OMR.  Perhaps it should if it is generally useful to eventual non-OpenJ9 relocatable compiles.

dsouzai commented:
Huh, I thought when I had added this in J9, I had also added the default OMR implementation. Clearly not...

In OpenJ9, the heuristic region is only used for relocatable compiles, to tell the infra that validation records are not needed. However it could be expand for other uses as well. As such, it does very much feel like something that only makes sense at the project level. 

I'll add a default empty function in OMR so that if/when such logic makes sense to exist at the OMR level, it can be added there.

dsouzai commented:
> Perhaps it should if it is generally useful to eventual non-OpenJ9 relocatable compiles.

Sorry, I missed the point specifically about relocatable compiles. The heuristic region is currently used to denote that validations are not needed. I don't believe generating validation records will ever make sense at the OMR level (because what requires validation very much depends on language semantics). So, I don't think deciding when validations are needed can exist at the OMR level either.

andrewcraik commented:
@dsouzai the enter-exit style of API has the possibility of being mismatched with no obvious warning that it has happened. Could the enter/exit be wrapped in a helper like the VMAccess regions so that the enter and exit are tied to a lexical scope? Might make the code more robust to future editing and would follow the convention established for VM access...

andrewcraik commented:
The heuristic region change itself seems reasonable, though I would prefer doxygen on the actual enter and exit functions so that they doxygen nicely.

dsouzai commented:
> @dsouzai the enter-exit style of API has the possibility of being mismatched with no obvious warning that it has happened. Could the enter/exit be wrapped in a helper like the VMAccess regions so that the enter and exit are tied to a lexical scope? Might make the code more robust to future editing and would follow the convention established for VM access...

Yeah I can add a `TR::HeuristicRegion` RAII wrapper.

> The heuristic region change itself seems reasonable, though I would prefer doxygen on the actual enter and exit functions so that they doxygen nicely.

Sure will do.

dsouzai commented:
Requested changes made.

vijaysun-omr commented:
@andrewcraik please review and give approval if your comments have been addressed.

vijaysun-omr commented:
@genie-omr build all

vijaysun-omr commented:
Looks reasonable to me. 

I personally worry more about the lack of explicit enter/exit since the exit is done implicitly in the current code when the live range of the heuristic range ends. e.g. if this kind of pattern is used elsewhere and some access that cannot be ignored happened near the code that did an access that can be ignored accidentally. But really a bug such as that can even happen with an enter/exit if one is not careful, but it's just a bit more obvious what's happening with an explicit enter/exit. 

Having said all that, the explicit enter/exit api does exist and it is safe to have the code as it is in the specific spot you changed.

vijaysun-omr commented:
Checks have passed. Merging.
